src/actions/tamanu/upgrade.rs

```diff
@@ -0,0 +1,192 @@
+use std::{fs, path::PathBuf};
+
+use clap::Parser;
+use miette::{bail, miette, Context as _, IntoDiagnostic, Result};
+use node_semver::Version;
+use regex::RegexBuilder;
+use tracing::info;
+
+use crate::actions::{caddy::configure_tamanu::DEFAULT_CADDYFILE_PATH, Context};
+
+use super::{
+	find_package, find_tamanu, prepare_upgrade::UPGRADED_SIGNAL_NAME, ApiServerKind, TamanuArgs,
+};
+
+/// Perform an upgrade.
+///
+/// This will incur downtime.
+///
+/// This command will detect which server is installed (Facility or Central) and which version is
+/// currently running, then download the desired newer version, unpack it, copy config across,
+/// install dependencies, and perform readiness checks.
+/// This command will checks if ready to upgrade, updates caddy configuration to new web folder,
+/// switches caddy to upgrade mode (always return 526 code for external connections),
+/// take down running API servers, run migrations, start new API servers, does healthcheck, switch caddy back to normal mode
```

Comment:
```suggestion
/// Perform an upgrade.
///
/// This will incur downtime.
///
/// `bestool tamanu pre-upgrade` must be run before this command.
///
/// This command will switch the server to upgrade mode, take down API servers, migrate the database,
/// start new API servers, upgrade the frontend, healthcheck, and finally switch out of upgrade mode.
```